designs/RestoreAndBuildCompatibility.md

```diff
@@ -0,0 +1,224 @@
+
+# Improving the restore and build compatibility experience
+
+* Status: In Review
+* Author(s): [Nikolche Kolev](https://github.com/nkolev92)
+* Issue: [10259](https://github.com/NuGet/Home/issues/10259) Improving the experience when incompatible versions of the restore/build tooling are used
+
+## Problem Background
+
+### NuGet packages.config
+
+NuGet as a product has grown its importance in the .NET ecosystem over the years.
+What started with NuGet.exe & Visual Studio support has led to a first class experience in **the CLI**, dotnet.exe.
+
+With packages.config, NuGet was built on top of the build experience, NuGet functions as an add on.
+NuGet used concept such as `Reference` to add package dependencies. These references were written in the project file directly, with a relative path to a `known` packages folder.
+This meant that you can use NuGet.exe 3.4.4 with VS 2015 & VS 2013 as the build tooling.
+
+The workflow for that usually is:
+
+* NuGet.exe restore - which downloads a defined list of packages to a specific folder.
+* build - with your tooling of choice. Said tooling doesn't need to be aware of the fact that NuGet is involved in any way.
+
+### NuGet PackageReference
+
+Since the introduction of project.json and PackageReference (2015 & 2017 respectively), NuGet has become tightly integrated with the build.
+Today NuGet's PackageReference is the tooling of choice for *all* .NET Core, now one unified .NET, projects.
+
+With PackageReference projects a restore is now *a must*.
+In addition to `NuGet.exe`, you can restore PackageReference based projects with `msbuild /t:restore` and `dotnet.exe restore`.
+
+The `restore` and `build` operations have a contract expressed through 3 files:
+
+* `project.assets.json`
+* `projectName.ext.nuget.g.props`
+* `projectName.ext.nuget.g.targets`
+
+At build-time these files are loaded and interpreted as necessary.
+This means that the NuGet and build tooling version have an implied compatibility matrix.
+
+### Restore, Build and Continue Integration
+
+When developing in Visual Studio or any other editor, the restore and build steps are integrated and things normally just work.  
+
+On the CI, the story is a little bit different.
+
+#### Tooling options
+
+Starting with VS 2017 and now VS 2019, there 3 different tooling options on can use to run a restore on their CI.
+
+* MSBuild.exe
+  * Ships with Visual Studio
+  * Supports PackageReference
+  * Starting with VS 2019, Update 5, packages.config scenarios in an opt-in fashion through -p:RestorePackagesConfig=true`

```

Comment:
I would recommend putting RestorePackagesConfig in a Directory.Build.props file instead of having to pass it on the command line.

Reply:
added.